Go docs update (#130)
undefinedlabs/scope-docs: docs/go-configuration.md

sidebar_label: Configuration
---

The behaviour of Scope Agent can be modified either by code or using environment variables.

Configuration by code is supported for both `scopeagent.Run(m)` and `agent.NewAgent()` as agent options parameters.

> If these properties are manually configured, use `true` or `false` only for the boolean value.

## Changing service name

You can specify the name of the service when sending data to Scope. Scope uses `default` as fallback.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_SERVICE="service-name"
```

> <!--By Code-->

Reviewer: ```suggestion
<!--Runtime instrumentation-->
```

and I'd also add the example in `Testing instrumentation` using `scopeagent.Run(m)`

Author: Ok

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_TRACER_GLOBAL=true
```

<!--By Code-->

```go
import (
    "go.undefinedlabs.com/scopeagent/agent"
)

func main() {
    scopeAgent, err := agent.NewAgent(agent.WithSetGlobalTracer())
    if err != nil {
        panic(err)
    }
    defer scopeAgent.Stop()

    // ...
}
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Changing testing mode

This variable indicates whether the agent is running tests or it is being used for runtime instrumentation. 

If `SCOPE_TESTING_MODE` is not specified, the agent sets to  `true` when running in CI or if the tests are started using `scopeagent.Run()` or if is instrumented using the `go.undefinedlabs.com/scopeagent/autoinstrument` import,  and `false` otherwise.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_TESTING_MODE=false
```

<!--By Code-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithTestingModeEnabled()))
}

// ...
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Enabling test retries on failure

> The Scope agent has support for execution retries on failing tests. By default, no retries are configured.

Reviewer: ```suggestion
The Scope agent has support for retrying failed tests. By default, no retries are performed.
```

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_TESTING_FAIL_RETRIES=3
```

<!--By Code-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithRetriesOnFail(3)))
}

// ...
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Handling test panic as a failure

> By default, the standard testing framework exits the test process on panic. This behavior can be changed to handle panics as fails, allowing the complete execution of the test suite. 

Reviewer: ```suggestion
By default, the standard testing framework exits the test process on panic. This behavior can be changed to handle panics as test failures, allowing the complete execution of the test suite. 
```

Also, this option can be combined with test retries, to retry tests that panic.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_TESTING_PANIC_AS_FAIL=true
```

<!--By Code-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithHandlePanicAsFail()))
}

// ...
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Adding agent metadata

> You can send arbitrary metadata for every test associated with certain commit which will be shown by Scope.

Reviewer: ```suggestion
You can add custom agent metadata that will be appended to the default metadata that the agent collects (such as host information).
```

Additionally, it is possible to set an environment variable as metadata value for a certain key, whose final value will be evaluated at runtime.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_METADATA="CustomKey1=MyValue, CustomKey2=MyValue2, CustomPath=\$PATH"
```

This is equivalent to: 
```go
    map[string]interface{}{
		"CustomKey1": "MyValue",
		"CustomKey2": "MyValue2",
		"CustomPath": os.Getenv("PATH"),
	}
```

<!--By Code-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithMetadata(map[string]interface{}{
		"CustomKey1": "MyValue",
		"CustomKey2": "MyValue2",
		"CustomPath": os.Getenv("PATH"),
	})))
}

// ...
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Select agent metadata as test configuration

> Yon can select metadata keys to be considered as relevant in the configuration of the test.

Reviewer: For the future, we should have a section that explains what a _configuration_ is.

Author: ok

    if err != nil {
        panic(err)
    }
    defer scopeAgent.Stop()

    // ...
}
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Changing testing mode

This variable indicates whether the agent is running tests or it is being used for runtime instrumentation. 

If `SCOPE_TESTING_MODE` is not specified, the agent sets to  `true` when running in CI or if the tests are started using `scopeagent.Run()` or if is instrumented using the `go.undefinedlabs.com/scopeagent/autoinstrument` import,  and `false` otherwise.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_TESTING_MODE=false
```

<!--Testing instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithTestingModeEnabled()))
}

func TestExample(t *testing.T) {
    // ...
}

func BenchmarkExample(b *testing.B) {
    // ...
}
```

<!--Runtime instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {

>     os.Exit(scopeagent.Run(m, agent.WithTestingModeEnabled()))

Reviewer: This doesn't seem like runtime instrumentation

Author: ok, my bad

}

// ...
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Enabling test retries on failure

The Scope agent has support for retrying failed tests. By default, no retries are performed.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_TESTING_FAIL_RETRIES=3
```

<!--Testing instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithRetriesOnFail(3)))
}

func TestExample(t *testing.T) {
    // ...
}

func BenchmarkExample(b *testing.B) {
    // ...
}
```

<!--Runtime instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {

>     os.Exit(scopeagent.Run(m, agent.WithRetriesOnFail(3)))

Reviewer: This doesn't seem like runtime instrumentation

Author: ok, my bad

}

// ...
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Handling test panic as a failure

By default, the standard testing framework exits the test process on panic. This behavior can be changed to handle panics as test failures, allowing the complete execution of the test suite. 

Also, this option can be combined with test retries, to retry tests that panic.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_TESTING_PANIC_AS_FAIL=true
```

<!--Testing instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithHandlePanicAsFail()))
}

func TestExample(t *testing.T) {
    // ...
}

func BenchmarkExample(b *testing.B) {
    // ...
}
```

<!--Runtime instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {

>     os.Exit(scopeagent.Run(m, agent.WithHandlePanicAsFail()))

Reviewer: This doesn't seem like runtime instrumentation

Author: ok

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_METADATA="CustomKey1=MyValue, CustomKey2=MyValue2, CustomPath=\$PATH"
```

This is equivalent to: 
```go
    map[string]interface{}{
		"CustomKey1": "MyValue",
		"CustomKey2": "MyValue2",
		"CustomPath": os.Getenv("PATH"),
	}
```

<!--Testing instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithMetadata(map[string]interface{}{
        "CustomKey1": "MyValue",
        "CustomKey2": "MyValue2",
        "CustomPath": os.Getenv("PATH"),
    })))
}

func TestExample(t *testing.T) {
    // ...
}

func BenchmarkExample(b *testing.B) {
    // ...
}
```

<!--Runtime instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithMetadata(map[string]interface{}{
		"CustomKey1": "MyValue",
		"CustomKey2": "MyValue2",
		"CustomPath": os.Getenv("PATH"),

> 	})))

Reviewer: This doesn't seem like runtime instrumentation

Author: ok

Yon can select metadata keys to be considered as relevant in the configuration of the test.

By default, Scope Go agent adds the folowwing metadata keys to the test configuration:

- **platform.name**: the result of calling `runtime.GOOS`
- **platform.architecture**: depends of the current architecture, can be `X64`, `X86`, `Arm`, `Arm64`
- **go.version**: the result of calling `runtime.Version()`

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_CONFIGURATION="CustomKey1, CustomKey2"
```

This is equivalent to: 
```go
    []string{"CustomKey1", "CustomKey2"}
```

<!--Testing instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {
    os.Exit(scopeagent.Run(m, agent.WithConfigurationKeys([]string{"CustomKey1", "CustomKey2"})))
}

func TestExample(t *testing.T) {
    // ...
}

func BenchmarkExample(b *testing.B) {
    // ...
}
```

<!--Runtime instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/agent"
)

func TestMain(m *testing.M) {

>     os.Exit(scopeagent.Run(m, agent.WithConfigurationKeys([]string{"CustomKey1", "CustomKey2"})))

Reviewer: This doesn't seem like runtime instrumentation

Author: ok

By default, Scope Go Agent will not send HTTP payloads as they are considered sensitive information.

<!--DOCUSAURUS_CODE_TABS-->
<!--Environment Variable-->

```sh
SCOPE_INSTRUMENTATION_HTTP_PAYLOADS=true
```

<!--Testing instrumentation-->

```go
import (
    "os"
    "testing"

    "go.undefinedlabs.com/scopeagent"
    "go.undefinedlabs.com/scopeagent/instrumentation/nethttp"
)

func TestMain(m *testing.M) {
    // Default Http Client
	nethttp.PatchHttpDefaultClient(nethttp.WithPayloadInstrumentation())

    os.Exit(scopeagent.Run(m))
}

func TestExample(t *testing.T) {
    // Custom http.Client instance
    client := &http.Client{Transport: &nethttp.Transport{ PayloadInstrumentation: true }}

    // ...
}

func BenchmarkExample(b *testing.B) {
    // ...
}
```

<!--Runtime instrumentation-->

```go
import (
    "go.undefinedlabs.com/scopeagent/instrumentation/nethttp"
)

func main() {
    // Default Http Client
    nethttp.PatchHttpDefaultClient(nethttp.WithPayloadInstrumentation())

    // Custom http.Client instance
    client := &http.Client{Transport: &nethttp.Transport{ PayloadInstrumentation: true }}

    // ...
}
```

<!--END_DOCUSAURUS_CODE_TABS-->

> ## Changing the log folder

Reviewer: ```suggestion
## Changing the log directory
```

> By default, the log folder were the Scope agent stores all the execution logs depends on the platform running:

Reviewer: ```suggestion
By default, the log directory were the Scope agent stores its logs depends on the platform:
```

```
Linux = /var/log/scope
Windows = ${HOME}/AppData/Roaming/scope/logs
MacOs = ${HOME}/Library/Logs/Scope

> ```

Reviewer: Not sure why this is a code block. Just do a markdown list instead:

```md
* Linux: `/var/log/scope`
* Windows: `${HOME}/AppData/Roaming/scope/logs`
* Mac OS: `${HOME}/Library/Logs/Scope`

Author: ok

> > If the agent can't write to the log folder, then will ask the OS for a temp folder as: `${TMP}/scope`

Reviewer: ```suggestion
> If the agent can't write to the default log folder, then it will create a temporary directory
```

> You can overwrite and specify a custom folder were the Scope agent will store the execution logs.

Reviewer: ```suggestion
You can also specify a custom log directory:
```

```sh
SCOPE_LOGGER_ROOT=/home/user/my-logs
```

## Disabling monkey patching

> To enable some auto instrumentations, the Scope agent uses monkey patching over functions of the standard `testing` package.

Reviewer: ```suggestion
To autoinstrument tests, the Scope agent uses an experimental monkey patching technique, which can cause issues in some cases. You can decide to disable monkey patching, but note the following:
```
And remove the next line

By disabling monkey patching the following features are disabled:

> - `testing.T` logs instrumentation: as an alternative, you can use the log methods from the scope test struct `scopeagent.GetTest(t)` as explained in the Logging Instrumentation section.

Reviewer: ```suggestion
- `testing.T` test logs are not automatically instrumented. As an alternative, you can use the log methods from the scope test struct `scopeagent.GetTest(t)` as explained in the [Logging Instrumentation](go-logging-instrumentation.md) section.
```

- Tests auto instrumentation using the `go.undefinedlabs.com/scopeagent/autoinstrument` import: as an alternative, you will have to create a `TestMain` as explained in the Installation section.

To disable monkey patching:

```sh
SCOPE_DISABLE_MONKEY_PATCHING=true

> ```

Reviewer: Question: why don't we patch the testing logger directly on `autoinstrument`, so there's no need for this env var, if you don't want monkey patching, you just don't import `autoinstrument`?

Author: Agree, that's simpler for the user... I'm gonna do a `PR` to do that refactor.
